### ISSUE TITLE
Question on the modeling unit set

### ISSUE THREAD

**Issue by wenjie-p** [espnet/espnet#3053]
Hi,

Thanks for the great toolkit you have built! I have been using this toolkit for a while but got confused on how `espnet` deal with rare units.

It looks to me the modeling units set generated by `espnet` is through using `sentencepiece` to encode the `training` text, where most words can be encoded into normal `piece`, while the rare units are reserved in the `dict`. (suggested in this [line of code)](https://github.com/espnet/espnet/blob/master/egs/commonvoice/asr1/run.sh#L149). Although `<unk>` existed in the dict set, it is not used for the rare units to map. In addition, the `<unk>` seems never appear in training process since all the `rare units` in the training set are reserved. Maybe it will be more reasonable to map the rare units into `<unk>` for model training?

I am not sure whether my understanding is correct or not, looking for your reply!

**Comment by sw005320** [comment]
The use of sentencepiece is actually to avoid <unk>.
All rare words can be a compound word of the piece and we can still handle them without introducing <unk>.
The reason we still maintain <unk> is that we have some unseen graphemes in the test set (but this is just used for logging. Not used for the actual algorithm) or some advanced decoding method (like word-based LM) requires <unk>.

**Comment by wenjie-p** [comment]
> The use of sentencepiece is actually to avoid .

Do you mean the use of `sentencepiece` is just providing tokenized results, the need of `sentencepiece` should be minimized for `espnet`?

> All rare words can be a compound word of the piece and we can still handle them without introducing .

In my observation, the words involved rare units can be divided into two cases: 1) part of the word are rare units; 2) the whole word can be rare units. I think what you mean is the first case, where we still have the potential to recognize part of the word? Please correct me if my understanding is wrong.

> The reason we still maintain is that we have some unseen graphemes in the test set (but this is just used for logging. Not used for the actual algorithm) or some advanced decoding method (like word-based LM) requires .

I think I can understand the use for logging since all the unseen rare units in test set will never be recognized, but I don't understand the reason `Not used for the actual algorithm`.  What you have mentioned is at the stage of inference, but rare units could have influence on model training. Training on many units with extremely low frequency (i.e. 3) should be harmful for the model.

**Comment by sw005320** [comment]
> > The use of sentencepiece is actually to avoid .
> 
> Do you mean the use of `sentencepiece` is just providing tokenized results, the need of `sentencepiece` should be minimized for `espnet`?
> 
> > All rare words can be a compound word of the piece and we can still handle them without introducing .
> 
> In my observation, the words involved rare units can be divided into two cases: 1) part of the word are rare units; 2) the whole word can be rare units. I think what you mean is the first case, where we still have the potential to recognize part of the word. Please correct me if my understanding is wrong.
> 
> > The reason we still maintain is that we have some unseen graphemes in the test set (but this is just used for logging. Not used for the actual algorithm) or some advanced decoding method (like word-based LM) requires .
> 
> I think I can understand the use for logging since all the unseen rare units in test set will never be recognized, but I don't understand the reason `Not used for the actual algorithm`. What you have mentioned is at the stage of inference, but rare units could have influence on model training. Training on many units with extremely low frequency (i.e. 3) should be harmful for the model.

Sorry, I cannot understand your questions.
Anyway, we usually using top N units, and rare units (low-frequency ones) are not used at all.
So, as long as we set N correctly, we would not have serious low-frequency issues.

**Comment by wenjie-p** [comment]
I am very sorry for not making it clear. I think I should give some concrete details.

I am running the recipe of egs/commonvoice/asr1 on the German data part. In the run.sh, the bpe unit size is set to 150, while the output size is 346 as shown in the model.json. By going through the units set in the model.json, I found many units only appears 3 times in the training transcripts. I also checked the text in the dump dir for model training, where the rare units is kept for model training. I also found although <unk> is kept in the units set, it seems not to be trained because I didn't found the unk_id in the transcripts under dump dir. So I am wondering why didn't map rare units into <unk> for model training, because this mapping may be helpful for the model performance I think.
